asylo/platform/primitives/fortanix_edp/edp_runner/src/lib.rs

```diff
@@ -0,0 +1,258 @@
+                Ok(())
+            },
+            Err(e) => Err(e)
+        }
+    })())
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn fortanix_edp_free_output_buffer(output: *mut c_uchar, output_size: usize) {
+    if !output.is_null() {
+        let _ = Box::from_raw(slice::from_raw_parts_mut(output, output_size));
+    }
+}
+
+#[repr(C)]
+pub struct EdpEnclave {
+    user_to_enclave_tx: Sender<Vec<u8>>,
+    enclave_to_user_rx: Receiver<Vec<u8>>,
+    main_thread: thread::JoinHandle<()>,
+}
+
+impl EdpEnclave {
+    fn load_enclave(path: &str) -> Result<Self, Error> {
+        let mut device = IsgxDevice::new()
+            .context("While opening SGX device")?
+            .einittoken_provider(AesmClient::new())
+            .build();
+        let mut enclave_builder = EnclaveBuilder::new(path.as_ref());
+        enclave_builder.with_dummy_signature_signer::<Hasher, _, _, _, _>(|der| {
+            PrivateKey::from_private_key_der(der).unwrap()
+        });
+
+        let (user_to_enclave_tx, user_to_enclave_rx) = channel();
+        let (enclave_to_user_tx, enclave_to_user_rx) = channel();
+        let asylo_extension = AsyloExtension {
+            inner: Mutex::new(Some(MessageStream::new(enclave_to_user_tx, user_to_enclave_rx))),
+        };
+        enclave_builder.usercall_extension(asylo_extension);
+
+        let enclave = enclave_builder.build(&mut device).context("While loading SGX enclave")?;
+        let main_thread = thread::spawn(move || {
+            enclave.run().map_err(|e| {
+                println!("Error while executing SGX enclave.\n{}", e);
+                std::process::exit(1);
+            }).unwrap();
+        });
+        Ok(Self { user_to_enclave_tx, enclave_to_user_rx, main_thread })
+    }
+
+    fn enclave_call(&self, selector: u64, input: &[u8]) -> Result<Vec<u8>, Error> {
+        let mut msg = Vec::with_capacity(16 + input.len());
+        msg.extend_from_slice(&selector.to_be_bytes());
+        msg.extend_from_slice(&input.len().to_be_bytes());
+        msg.extend_from_slice(input);
+        self.user_to_enclave_tx.send(msg)?;
+
+        let mut resp = Vec::new();
+        let mut size = None;
+        loop {
+            let mut v = self.enclave_to_user_rx.recv()?;
+            resp.append(&mut v);
+            if size.is_none() && resp.len() >= mem::size_of::<usize>() {
+                let mut s = [0u8; mem::size_of::<usize>()];
+                s.copy_from_slice(&resp[0..mem::size_of::<usize>()]);
+                let s = usize::from_be_bytes(s);
+                size = Some(s);
+            }
+            if size.map_or(false, |size| resp.len() >= mem::size_of::<usize>() + size) {
+                break;
+            }
+        }
+        let size = size.unwrap();
+        // extra data received here would not be accessible later
+        assert_eq!(resp.len(), mem::size_of::<usize>() + size);
+        Ok(resp[mem::size_of::<usize>()..mem::size_of::<usize>() + size].to_owned())
```

Comment:
This is pretty inefficient. Can you read the length first in a separate buffer and then prepare the to-be-returned output buffer directly?

Reply:
refactored to also avoid throwing away extra bytes read